// devextreme-angular/esm2015/core/transfer-state.js
/*!
 * devextreme-angular
 * Version: 23.2.11
 * Build date: Mon Dec 16 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import { PLATFORM_ID, Inject, NgModule } from '@angular/core';
import { isPlatformServer } from '@angular/common';
import ajax from 'devextreme/core/utils/ajax';
import { Deferred } from 'devextreme/core/utils/deferred';
import { TransferState, makeStateKey } from '@angular/platform-browser';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class DxServerTransferStateModule {
    constructor(state, platformId) {
        this.state = state;
        this.platformId = platformId;
        const that = this;
        ajax.inject({
            sendRequest(...args) {
                const key = makeStateKey(that.generateKey(args));
                const cachedData = that.state.get(key, null);
                if (isPlatformServer(that.platformId)) {
                    const result = this.callBase.apply(this, args);
                    result.always((data, status) => {
                        const dataForCache = {
                            data,
                            status,
                        };
                        that.state.set(key, dataForCache);
                    });
                    return result;
                }
                if (cachedData) {
                    const d = Deferred();
                    d.resolve(cachedData.data, cachedData.status);
                    that.state.set(key, null);
                    return d.promise();
                }
                return this.callBase.apply(this, args);
            },
        });
    }
    generateKey(args) {
        let keyValue = '';
        for (const key in args) {
            if (typeof args[key] === 'object') {
                const objKey = this.generateKey(args[key]);
                keyValue += key + objKey;
            }
            else {
                keyValue += key + args[key];
            }
        }
        return keyValue;
    }
}
/** @nocollapse */ DxServerTransferStateModule.ɵfac = function DxServerTransferStateModule_Factory(t) { return new (t || DxServerTransferStateModule)(i0.ɵɵinject(i1.TransferState), i0.ɵɵinject(PLATFORM_ID)); };
/** @nocollapse */ DxServerTransferStateModule.ɵmod = /** @pureOrBreakMyCode */ i0.ɵɵdefineNgModule({ type: DxServerTransferStateModule });
/** @nocollapse */ DxServerTransferStateModule.ɵinj = /** @pureOrBreakMyCode */ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxServerTransferStateModule, [{
        type: NgModule,
        args: [{}]
    }], function () { return [{ type: i1.TransferState }, { type: undefined, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, null); })();
