// devextreme/esm/__internal/scheduler/r1/components/timeline/header_panel_timeline.js
/**
 * DevExtreme (esm/__internal/scheduler/r1/components/timeline/header_panel_timeline.js)
 * Version: 23.2.11
 * Build date: Mon Dec 16 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import {
    createComponentVNode
} from "inferno";
import {
    createReRenderEffect,
    InfernoWrapperComponent
} from "@devextreme/runtime/inferno";
import {
    getTemplate
} from "../../../../core/r1/utils/index";
import {
    HeaderPanel,
    HeaderPanelDefaultProps
} from "../base/header_panel";
import {
    TimelineDateHeaderLayout
} from "./date_header_timeline";
export class HeaderPanelTimeline extends InfernoWrapperComponent {
    createEffects() {
        return [createReRenderEffect()]
    }
    render() {
        var {
            viewContext: viewContext,
            dateCellTemplate: dateCellTemplate,
            dateHeaderData: dateHeaderData,
            groupByDate: groupByDate,
            groupOrientation: groupOrientation,
            groupPanelData: groupPanelData,
            groups: groups,
            isRenderDateHeader: isRenderDateHeader,
            resourceCellTemplate: resourceCellTemplate,
            timeCellTemplate: timeCellTemplate
        } = this.props;
        var DateCellTemplateComponent = getTemplate(dateCellTemplate);
        var ResourceCellTemplateComponent = getTemplate(resourceCellTemplate);
        var TimeCellTemplateComponent = getTemplate(timeCellTemplate);
        return createComponentVNode(2, HeaderPanel, {
            viewContext: viewContext,
            dateHeaderData: dateHeaderData,
            groupPanelData: groupPanelData,
            groupByDate: groupByDate,
            groups: groups,
            groupOrientation: groupOrientation,
            isRenderDateHeader: isRenderDateHeader,
            dateHeaderTemplate: TimelineDateHeaderLayout,
            resourceCellTemplate: ResourceCellTemplateComponent,
            dateCellTemplate: DateCellTemplateComponent,
            timeCellTemplate: TimeCellTemplateComponent
        })
    }
}
HeaderPanelTimeline.defaultProps = HeaderPanelDefaultProps;
