// devextreme-angular/esm2015/ui/nested/tasks.js
/*!
 * devextreme-angular
 * Version: 23.2.11
 * Build date: Mon Dec 16 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { NestedOption } from 'devextreme-angular/core';
import * as i0 from "@angular/core";
import * as i1 from "devextreme-angular/core";
export class DxoTasksComponent extends NestedOption {
    constructor(parentOptionHost, optionHost) {
        super();
        parentOptionHost.setNestedOption(this);
        optionHost.setHost(this, this._fullOptionPath.bind(this));
    }
    get colorExpr() {
        return this._getOption('colorExpr');
    }
    set colorExpr(value) {
        this._setOption('colorExpr', value);
    }
    get dataSource() {
        return this._getOption('dataSource');
    }
    set dataSource(value) {
        this._setOption('dataSource', value);
    }
    get endExpr() {
        return this._getOption('endExpr');
    }
    set endExpr(value) {
        this._setOption('endExpr', value);
    }
    get keyExpr() {
        return this._getOption('keyExpr');
    }
    set keyExpr(value) {
        this._setOption('keyExpr', value);
    }
    get parentIdExpr() {
        return this._getOption('parentIdExpr');
    }
    set parentIdExpr(value) {
        this._setOption('parentIdExpr', value);
    }
    get progressExpr() {
        return this._getOption('progressExpr');
    }
    set progressExpr(value) {
        this._setOption('progressExpr', value);
    }
    get startExpr() {
        return this._getOption('startExpr');
    }
    set startExpr(value) {
        this._setOption('startExpr', value);
    }
    get titleExpr() {
        return this._getOption('titleExpr');
    }
    set titleExpr(value) {
        this._setOption('titleExpr', value);
    }
    get _optionPath() {
        return 'tasks';
    }
    ngOnInit() {
        this._addRecreatedComponent();
    }
    ngOnDestroy() {
        this._addRemovedOption(this._getOptionPath());
    }
}
/** @nocollapse */ DxoTasksComponent.ɵfac = function DxoTasksComponent_Factory(t) { return new (t || DxoTasksComponent)(i0.ɵɵdirectiveInject(i1.NestedOptionHost, 5), i0.ɵɵdirectiveInject(i1.NestedOptionHost, 1)); };
/** @nocollapse */ DxoTasksComponent.ɵcmp = /** @pureOrBreakMyCode */ i0.ɵɵdefineComponent({ type: DxoTasksComponent, selectors: [["dxo-tasks"]], inputs: { colorExpr: "colorExpr", dataSource: "dataSource", endExpr: "endExpr", keyExpr: "keyExpr", parentIdExpr: "parentIdExpr", progressExpr: "progressExpr", startExpr: "startExpr", titleExpr: "titleExpr" }, features: [i0.ɵɵProvidersFeature([NestedOptionHost]), i0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxoTasksComponent_Template(rf, ctx) { }, styles: [""] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxoTasksComponent, [{
        type: Component,
        args: [{
                selector: 'dxo-tasks',
                template: '',
                styles: [''],
                providers: [NestedOptionHost]
            }]
    }], function () { return [{ type: i1.NestedOptionHost, decorators: [{
                type: SkipSelf
            }, {
                type: Host
            }] }, { type: i1.NestedOptionHost, decorators: [{
                type: Host
            }] }]; }, { colorExpr: [{
            type: Input
        }], dataSource: [{
            type: Input
        }], endExpr: [{
            type: Input
        }], keyExpr: [{
            type: Input
        }], parentIdExpr: [{
            type: Input
        }], progressExpr: [{
            type: Input
        }], startExpr: [{
            type: Input
        }], titleExpr: [{
            type: Input
        }] }); })();
export class DxoTasksModule {
}
/** @nocollapse */ DxoTasksModule.ɵfac = function DxoTasksModule_Factory(t) { return new (t || DxoTasksModule)(); };
/** @nocollapse */ DxoTasksModule.ɵmod = /** @pureOrBreakMyCode */ i0.ɵɵdefineNgModule({ type: DxoTasksModule });
/** @nocollapse */ DxoTasksModule.ɵinj = /** @pureOrBreakMyCode */ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxoTasksModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DxoTasksComponent
                ],
                exports: [
                    DxoTasksComponent
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(DxoTasksModule, { declarations: [DxoTasksComponent], exports: [DxoTasksComponent] }); })();
