// devextreme-angular/esm2015/ui/nested/validation.js
/*!
 * devextreme-angular
 * Version: 23.2.11
 * Build date: Mon Dec 16 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { NestedOption } from 'devextreme-angular/core';
import * as i0 from "@angular/core";
import * as i1 from "devextreme-angular/core";
export class DxoValidationComponent extends NestedOption {
    constructor(parentOptionHost, optionHost) {
        super();
        parentOptionHost.setNestedOption(this);
        optionHost.setHost(this, this._fullOptionPath.bind(this));
    }
    get autoUpdateParentTasks() {
        return this._getOption('autoUpdateParentTasks');
    }
    set autoUpdateParentTasks(value) {
        this._setOption('autoUpdateParentTasks', value);
    }
    get enablePredecessorGap() {
        return this._getOption('enablePredecessorGap');
    }
    set enablePredecessorGap(value) {
        this._setOption('enablePredecessorGap', value);
    }
    get validateDependencies() {
        return this._getOption('validateDependencies');
    }
    set validateDependencies(value) {
        this._setOption('validateDependencies', value);
    }
    get _optionPath() {
        return 'validation';
    }
    ngOnInit() {
        this._addRecreatedComponent();
    }
    ngOnDestroy() {
        this._addRemovedOption(this._getOptionPath());
    }
}
/** @nocollapse */ DxoValidationComponent.ɵfac = function DxoValidationComponent_Factory(t) { return new (t || DxoValidationComponent)(i0.ɵɵdirectiveInject(i1.NestedOptionHost, 5), i0.ɵɵdirectiveInject(i1.NestedOptionHost, 1)); };
/** @nocollapse */ DxoValidationComponent.ɵcmp = /** @pureOrBreakMyCode */ i0.ɵɵdefineComponent({ type: DxoValidationComponent, selectors: [["dxo-validation"]], inputs: { autoUpdateParentTasks: "autoUpdateParentTasks", enablePredecessorGap: "enablePredecessorGap", validateDependencies: "validateDependencies" }, features: [i0.ɵɵProvidersFeature([NestedOptionHost]), i0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxoValidationComponent_Template(rf, ctx) { }, styles: [""] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxoValidationComponent, [{
        type: Component,
        args: [{
                selector: 'dxo-validation',
                template: '',
                styles: [''],
                providers: [NestedOptionHost]
            }]
    }], function () { return [{ type: i1.NestedOptionHost, decorators: [{
                type: SkipSelf
            }, {
                type: Host
            }] }, { type: i1.NestedOptionHost, decorators: [{
                type: Host
            }] }]; }, { autoUpdateParentTasks: [{
            type: Input
        }], enablePredecessorGap: [{
            type: Input
        }], validateDependencies: [{
            type: Input
        }] }); })();
export class DxoValidationModule {
}
/** @nocollapse */ DxoValidationModule.ɵfac = function DxoValidationModule_Factory(t) { return new (t || DxoValidationModule)(); };
/** @nocollapse */ DxoValidationModule.ɵmod = /** @pureOrBreakMyCode */ i0.ɵɵdefineNgModule({ type: DxoValidationModule });
/** @nocollapse */ DxoValidationModule.ɵinj = /** @pureOrBreakMyCode */ i0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxoValidationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DxoValidationComponent
                ],
                exports: [
                    DxoValidationComponent
                ],
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(DxoValidationModule, { declarations: [DxoValidationComponent], exports: [DxoValidationComponent] }); })();
