// devextreme-angular/esm2015/ui/nested/base/viz-scale-break-dxi.js
/*!
 * devextreme-angular
 * Version: 23.2.11
 * Build date: Mon Dec 16 2024
 *
 * Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
/* tslint:disable:max-line-length */
import { CollectionNestedOption } from 'devextreme-angular/core';
import { Component, } from '@angular/core';
import * as i0 from "@angular/core";
export class DxiVizScaleBreak extends CollectionNestedOption {
    get endValue() {
        return this._getOption('endValue');
    }
    set endValue(value) {
        this._setOption('endValue', value);
    }
    get startValue() {
        return this._getOption('startValue');
    }
    set startValue(value) {
        this._setOption('startValue', value);
    }
}
/** @nocollapse */ DxiVizScaleBreak.ɵfac = /** @pureOrBreakMyCode */ function () { let ɵDxiVizScaleBreak_BaseFactory; return function DxiVizScaleBreak_Factory(t) { return (ɵDxiVizScaleBreak_BaseFactory || (ɵDxiVizScaleBreak_BaseFactory = i0.ɵɵgetInheritedFactory(DxiVizScaleBreak)))(t || DxiVizScaleBreak); }; }();
/** @nocollapse */ DxiVizScaleBreak.ɵcmp = /** @pureOrBreakMyCode */ i0.ɵɵdefineComponent({ type: DxiVizScaleBreak, selectors: [["ng-component"]], features: [i0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function DxiVizScaleBreak_Template(rf, ctx) { }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.ɵsetClassMetadata(DxiVizScaleBreak, [{
        type: Component,
        args: [{
                template: ''
            }]
    }], null, null); })();
